### ISSUE TITLE
[Docs] Formalize (in writing) the masking system

### ISSUE THREAD

**Issue by femtomc** [genjax-community/genjax#235]
GenJAX uses an interesting system for supporting branching over structure. This issue highlights this system, and proposes a documentation page to illustrate when this system is utilized internally:

## Background

It's helpful to understand what you can and cannot do with JAX with control flow:

* Branching over structure or static values at trace time with Python control flow: allowed.
* Branching over structure at runtime with Python control flow: disallowed (structure is removed at runtime), Python control flow is unrolled at trace time -- if JAX finds a traced array in a branch condition, it will yell at you.
* Branching using array values at runtime: allowed - but strict requirements on branch bodies.

In general, you can use arbitrary control flow at trace time - as long as it can be statically resolved during tracing. But that doesn't provide us with a way to use control flow at runtime, or even write functions which support return values which are sum types (e.g. branch A returns type T_A, branch B returns type T_B).

For generative functions, we aspire to support expressive patterns of computation - including branches which support different choice map types. _This is inherently at odds with JAX (and XLA's) restrictions_ - thus, to encode this possibility, a bit of hacking is required.

## Enter `Sumtree`

GenJAX allows the construction of "sum types over Pytrees" using a special data structure called `Sumtree`:

```python
@dataclass
class Sumtree(Pytree):
    visitations: StaticCollection
    forms: StaticCollection
    payload: HashableDict
```

(The internals of the above might change, take with grain of salt)

The idea of `Sumtree` is simple: it's a tagged union which efficiently shares data across all of the `Pytree` data types in the union. The `PyTreeDef` forms of these types are stored in `Sumtree.forms`.  To create a `Sumtree`, you use the `new` constructor:

```python
@classmethod
def new(cls, source: Pytree, covers: Sequence[Pytree]):
    pass
```

Here, the "true value" is `source`. The point of this type is that you can construct `jax.lax.switch` branch bodies **at tracing time** whose "true data return types" are different, but they can be coerced into a consistent `Sumtree` type.

Said another way, the original JAX pattern:

```python
if A -> B else A -> B
```

with `Sumtree`, we can write:

```python
if A -> TUnion{B, C} else A -> TUnion{B, C}
```

This is useful for us -- it allows us to return different choice map / trace data types from `jax.lax.switch`, supporting the implementation of `Switch` combinator. These different types are coerced into the same tagged union type.

## Using `Sumtree` for `SwitchChoiceMap`

The problem is that the "tag" in tagged union is a JAX dynamic value (it's the index in `jax.lax.switch`, for example). So we can't immediately collapse the `Sumtree` outside the switch.

When we are inside the JAX API boundary (e.g. inside code which is intended to run on device), if we attempt to index or project the tagged union to the true value, we can't actually collapse it (because you can't switch on structure using a dynamically known value).

**Dynamic values raise new novel requirements on implementing generative functions - dynamic values arise because we want to support switching over structure**.

TODO: McCoy goes to class, returns and finishes.

**Comment by femtomc** [comment]
@derifatives Let's iterate on the masking system design here, on this issue.

The original issue post discussed why `Sumtree` was introduced, but below -- I'll discuss the `genjax.Mask` system.

**Comment by sritchie** [comment]
Migrated to internal Notion.
